### ISSUE TITLE
Very large VE may push WAL message beyond size limit

### ISSUE THREAD

**Issue by lasarojc** [cometbft/cometbft#1253]
WAL limits messages to 1MB.
Very large VE will push PreCommit messages beyond this limit.
This behavior should be captured in a specific UT and documented in the ABCI spec.

**Comment by adizere** [comment]
Should we consider upper bounding the side of VE to a (common sense) limit, eg 1MB? Comet is not a video streaming service so not sure what use-case could require large VE.

**Comment by sergio-mena** [comment]
If we decide to bound the size of VE, then it should probably be a consensus param (similar to the block size), with a hard-coded common-sense maximum size, like the block size

**Comment by adizere** [comment]
What is the behavior is a node tries to push into WAL a message that is > 1MB? Trying to assess how severe this problem may be.

**Comment by lasarojc** [comment]
Here is where it is tested
https://github.com/cometbft/cometbft/blob/c3f7ea35c12ff0ec3fc67b9f686a0da566dad73f/internal/consensus/wal.go#L319

> What is the behavior if a node tries to push into WAL a message that is > 1MB? Trying to assess how severe this problem may be.

This is actually curious. If it received from a peer, then log and go ahead. If from itself, then [panic](https://github.com/cometbft/cometbft/blob/c3f7ea35c12ff0ec3fc67b9f686a0da566dad73f/internal/consensus/state.go#L835)

**Comment by cason** [comment]
> This is actually curious. If it received from a peer, then log and go ahead. If from itself, then [panic](https://github.com/cometbft/cometbft/blob/c3f7ea35c12ff0ec3fc67b9f686a0da566dad73f/internal/consensus/state.go#L835)

We write asynchronously messages from peers, we fsync the WAL when we write or own message. The rationale here is probably that peers can send us invalid message, and we shouldn't break because of that, but we shouldn't broadcast (send to ourselves) invalid messages, so we panic.

**Comment by cason** [comment]
> What is the behavior if a node tries to push into WAL a message that is > 1MB? 

This is definitely problematic. While we should consider not writing blocks (split into parts or not) to the WAL, there is no possibility of non writing votes to the WAL. One possibility to consider is to not write vote extensions to the WAL. In particular because I don't see the use case for replaying vote extensions...

**Comment by andynog** [comment]
This is just to add more context to the problem. 

As Daniel said the problem is that with the `tendermint/wal/MsgInfo` written to the WAL which might be bigger than the allowed msg size in the WAL. 

Currently the WAL max message size is `maxMsgSizeBytes = maxMsgSize + 24` in `wal.go`, where `maxMsgSize = 1048576` (~1MB) in `consensus/reactor`

A `MsgInfo` on a WAL (converted to JSON) could looks like this:
```
{"time":"2024-04-08T21:05:11.089733Z","msg":{"type":"tendermint/wal/MsgInfo","value":{"msg":{"type":"tendermint/BlockPart","value":{"Height":"2","Round":0,"Part":{"index":0,"bytes":"0A80030A04080B10011211746573742D636861696E2D5237466E4E391802220B0886B8D1B00610B8859A232A480A20798D106E907255899C4DB68E6D7B0A1FB485E1698D28BDC12C1260FB739A108A12240801122049B359831AEEF15FDB179C4833F8750F41C685B68566D57065B61C1DA81066BB322081E6467AA5053DC485A10B1D50EDE4FF7ED5ADE86B3514A1A189047F8D7D59573A20E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B8554220BFC08C9E365339B52B02BA79F366C5E4C26B544D82E94811242CA49B1B16D82E4A20BFC08C9E365339B52B02BA79F366C5E4C26B544D82E94811242CA49B1B16D82E522068ECD6F333119CE43751ECE583B981F23508AEAF4221FF582B1BB33BE42BCEFA5A0800000000000000006220E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B8556A20E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B85572144C1949EDFEC2EB7AE430D50FB16706D84B27793E12001A0022B50108011A480A20798D106E907255899C4DB68E6D7B0A1FB485E1698D28BDC12C1260FB739A108A12240801122049B359831AEEF15FDB179C4833F8750F41C685B68566D57065B61C1DA81066BB2267080212144C1949EDFEC2EB7AE430D50FB16706D84B27793E1A0B0886B8D1B00610B8859A232240EE2CDDA334425AED705FFCA422007ADBF3442FEB5F00CBD2ABB87885D2B7AA6E7111E99CB5E80AEB0F1F5A2C2605E36908D1AD198C604C2539F8264D339AE806","proof":{"total":"1","index":"0","leaf_hash":"Or35gnue5Jn8uNxu0m8pxR8X2/THmec/feyufZDBqGY="}}}},"peer_key":"","receive_time":"0001-01-01T00:00:00Z"}}}```
```

So after some investigation, there are two methods in the WAL, `Write` and `WriteAsync`. In `WriteAsync` internally it invokes `Write`. Then in the `Write` method it invokes `Encode`.  In the `Encode` message it will the consensus message is converted to its equivalent proto message (NewValidBlock, BlockPart, VoteMessage, etc). 

Then in the encode there's a check for the length of the message to see if it is bigger than `maxMsgSizeBytes`

```
	data, err := proto.Marshal(&pv)
	if err != nil {
		panic(fmt.Errorf("encode timed wall message failure: %w", err))
	}

	crc := crc32.Checksum(data, crc32c)
	length := uint32(len(data))
	if length > maxMsgSizeBytes {
		return fmt.Errorf("msg is too big: %d bytes, max: %d bytes", length, maxMsgSizeBytes)
	}
```

if so, an error is returned. In most places where the error from `Write` is returned, just an error is logged and the logic continues. But in the `WriteSync` (as Lasaro) pointed out, it will panic

```
	case mi = <-cs.internalMsgQueue:
			err := cs.wal.WriteSync(mi) // NOTE: fsync
			if err != nil {
				panic(fmt.Sprintf(
					"failed to write %v msg to consensus WAL due to %v; check your file system and restart the node",
					mi, err,
				))
			}
```

while in the case of a peer message it would just log the error and continue

```
	case mi = <-cs.peerMsgQueue:
			if err := cs.wal.Write(mi); err != nil {
				cs.Logger.Error("failed writing to WAL", "err", err)
			}
			// handles proposals, block parts, votes
			// may generate internal events (votes, complete proposals, 2/3 majorities)
			cs.handleMsg(mi)
```

 So we need to ensure that the

**Comment by cason** [comment]
> Don't panic in WriteSync just return an error (what are the implications?)

We can proceed (namely, broadcast and process a message) if we are not able to sync the WAL. Otherwise, when recovering from a crash, we will "forget" that we sent that message (and that we received all messages preceding it, and therefore being a precondition for its production).

**Comment by andynog** [comment]
Thanks for the feedback @cason. So if it's possible to change the logic and not write the vote extensions to the WAL and no major implication would occur that might solve this problem as I understand. 

But the issue related with the size of vote extensions would still persist from a p2p perspective right ? Does the `p2p.max_packet_msg_payload_size` parameter ensures the ve size is capped from a P2P message perspective ?

**Comment by cason** [comment]
> Does the p2p.max_packet_msg_payload_size parameter

No, messages are split into packets. A message can take multiple packets. So this is not a problem. Recall that we ship application snapshots, blocks, etc., which are way larger.

**Comment by cason** [comment]
> So if it's possible to change the logic and not write the vote extensions to the WAL and no major implication would occur 

I am not 100% sure about that. @sergio-mena is the best to ask this question.

**Comment by sergio-mena** [comment]
Short answer: it MAY be possible, but it won't be a simple solution.

Long answer: As vote extensions are part of the precommit message, if we don't save them to the WAL, the recovered consensus instance will be missing them then we replay all consensus messages upon recovery.
This problem is similar to the one discussed in [RFC100](https://github.com/cometbft/cometbft/blob/main/docs/references/rfc/rfc-100-abci-vote-extension-propag.md) for nodes that are block-syncing, so a similar solution may exist.

**Comment by cason** [comment]
A good solution is far from simple. The main point is that vote messages were assumed to be fixed size, but now we add the ability of making them a variable-size message type. This has impacts in many ways, such as vote propagation, timeouts, and persistence.

Can we find a temporarily workaround, such as adding a maximum size for vote extensions?

**Comment by andynog** [comment]
Thanks @cason, I'm assuming this parameter would be a Consensus Parameter right ?

**Comment by cason** [comment]
It could only be hard-coded, for now, based on the maximum WAL message size.

**Comment by sergio-mena** [comment]
Remember that, in the current design, it is up to the application (typically as part of `VerifyVoteExtension`) to set (and check) the maximum size of vote extensions

**Comment by andynog** [comment]
This seems to be an application concern, I'll leave this open for future reference but feel free to close in case there's no immediate action.
